win32/tests/srw_lock_win32_ut/srw_lock_win32_ut.c

```diff
@@ -477,6 +595,114 @@ TEST_FUNCTION(srw_lock_acquire_shared_with_do_statistic_false_succeeds)
     srw_lock_destroy(bsdlLock);
 }
 
+/* srw_lock_try_acquire_shared */
+
+/* Tests_SRS_SRW_LOCK_01_001: [ If handle is NULL then srw_lock_try_acquire_shared shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(srw_lock_try_acquire_shared_with_handle_NULL_returns)
+{
+    ///arrange
+
+    ///act
+    int result = srw_lock_try_acquire_shared(NULL);
+
+    ///assert
+    ASSERT_ARE_NOT_EQUAL(int, 0, result);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+}
+
+/*Tests_SRS_SRW_LOCK_01_002: [ Otherwise srw_lock_try_acquire_shared shall call TryAcquireSRWLockShared. ]*/
+/*Tests_SRS_SRW_LOCK_01_004: [ If TryAcquireSRWLockShared returns TRUE, srw_lock_try_acquire_shared shall return 0. ]*/
+/*Tests_SRS_SRW_LOCK_01_005: [ If do_statistics is true and the timer created has recorded more than TIME_BETWEEN_STATISTICS_LOG seconds then statistics will be logged and the timer shall be started again. ]*/
+TEST_FUNCTION(srw_lock_try_acquire_shared_succeeds)
+{
+    ///arrange
+    SRW_LOCK_HANDLE bsdlLock = TEST_srw_lock_create(true, "test_lock");
+
+    STRICT_EXPECTED_CALL(mocked_TryAcquireSRWLockShared(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(timer_get_elapsed(IGNORED_ARG))
+        .SetReturn(0);
+
+    ///act
+    int result = srw_lock_try_acquire_shared(bsdlLock);
```

Comment:
>srw_lock_try_acquire_shared [](start = 17, length = 27)

 Ido have to ask why we need these APIs.

Because following this CR there's going to be another where they are used. So... a bit of look in the future, if you will. Just out of curiosity

Reply:
Ah, sure, sorry about that:
Currently when we do a change role we unwind all the operations that have failed (because they will fail due to NOT_PRIMARY).
We have set block epoch which literally takes a lock on the block to disallows new appends (appends acquire as shared, set block epoch acquires as exclusive).

the lock on the block is acquired exclusively when set block epoch starts and released exclusively when it completes.
So if the set block epoch starts, lock is acquired exclusively, append attempts to acquire shared and it blocks in there.
Then we have a NOT_PRIMARY and we want to unwind the operations, but unfortunately we are stuck in the acquire shared lock.

So the way I am thinking of fixing this is attempt to take the lock (if it fails bounce this up as busy). We do have a retry layer that will retry the operation if it fails so I am planning to simply reuse the retry in the bsi_user_retry layer.

---
In reply to: [461968384](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461968384) [](ancestors = 461968384)